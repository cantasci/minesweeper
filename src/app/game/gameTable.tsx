import React, { FC } from 'react';

import { makeStyles } from '@mui/styles';

import { GameClient } from './gameClient';
import GameTableCell from './gameTableCell';

export const useGameTableStyles = makeStyles({
  row: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
  },

  outer: {
    maxWidth: '100%',
    overflow: 'scroll',
  },
});

interface Props {
  isRunning: boolean;
  gameMap: string[];
}

export const GameTable: FC<Props> = ({ gameMap, isRunning }) => {
  const classes = useGameTableStyles();

  const onCellClick = (y: number, x: number) => {
    GameClient.socket.send(`open ${x} ${y}`);
  };

  const renderMap = () => {
    return gameMap.map((item: string, rowIndex: number) => {
      const squares = item.split('');
      const rowKey = `square-${rowIndex}`;

      const maxRows = gameMap.length;
      const maxColumns = squares.length;
      const row = squares.map((square: string, columnIndex: number) => {
        const cellKey = `square-${rowIndex}-${columnIndex}`;
        const testId = `square-${rowIndex}-${columnIndex}`;

        return (
          <GameTableCell
            key={cellKey}
            testId={testId}
            positions={{
              x: rowIndex,
              y: columnIndex,
              maxColumns,
              maxRows,
            }}
            disabled={!isRunning}
            content={square}
            onCellClick={onCellClick}
          />
        );
      });
      return (
        <div key={rowKey} className={classes.row}>
          {row}
        </div>
      );
    });
  };

  return <div className={classes.outer}>{renderMap()}</div>;
};
