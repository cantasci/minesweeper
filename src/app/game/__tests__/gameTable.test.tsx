import React from 'react';

import { cleanup } from '@testing-library/react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { GameTable } from '../gameTable';

describe('<GameTable />', () => {
  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });
  it('renders correctly and to match snapshot', () => {
    const gameMap: string[] = [];
    const tree = renderer.create(<GameTable isRunning gameMap={gameMap} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render empty state correctly', () => {
    const gameMap: string[] = [];
    const wrapper = mount(<GameTable isRunning={false} gameMap={gameMap} />);
    expect(wrapper.text().includes('')).toBe(true);
  });

  it('should find at least 1 square cell', () => {
    const gameMap: string[] = [
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
      '□□□□□□□□□□',
    ];
    const wrapper = mount(<GameTable isRunning gameMap={gameMap} />);
    const element = wrapper.find('square-1-1');
    expect(element).toBeTruthy();
  });
});
