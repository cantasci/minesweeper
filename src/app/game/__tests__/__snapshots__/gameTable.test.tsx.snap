// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameTable /> renders correctly and to match snapshot 1`] = `
<div
  className="makeStyles-outer-2"
/>
`;
